File: java/InterstellarCompVsComp.java

import java.util.Random;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
/**
 *
 * @author i14032
 */
public class InterstellarCompVsComp {

    private Random rd = new Random();
    private int numBH = rd.nextInt(3)+1;
    private int numWG = rd.nextInt(3)+1;
    private Board board = new Board(6, 5, 5, 2);

    private int xin = 0;
    private int xout = 0;
    private int yin = 0;
    private int yout = 0;

    /**
     *
     * @param isGanjil kalo masukin 0 buat bikin angka ganjil
     */
    private void setXIn(int isGanjil) {
        int x = rd.nextInt(2)+1;
        x+= isGanjil;
        if (x % 2 == 0) {
            x++;
        }
        this.xin = x;
    }

    /**
     *
     * @param isGanjil kalo masukin 0 buat bikin angka ganjil
     */
    private void setXOut(int isGanjil) {
        int x = rd.nextInt(2)+1;
        x+=isGanjil;
        if (x % 2 == 0) {
            x++;
        }
        this.xout = x;
    }

    /**
     * buat masukin nilai yin
     *
     * @param isGanjil kalo masukin 0 buat bikin angka ganjil
     */
    private void setYIn(int isGanjil) {
        int x = rd.nextInt(3) +1;
        x+=isGanjil;
        if (x % 2 == 0) {
            x++;
        }
        this.yin = x;
    }

    /**
     * buat masukin nilai yout
     *
     * @param isGanjil kalo masukin 0 buat bikin angka ganjil
     */
    private void setYOut(int isGanjil) {
        int x = rd.nextInt(3)+1;
        x+=isGanjil;
        if (x % 2 == 0) {
            x++;
        }
        this.yout = x;
    }

    /**
     * buat objek blackhole sesuai jumlah blackhole
     */
    public void makeAllBH() {
        //blackhole di ganjil
        if(numBH%2!=0){
             numBH++;
        }
        for (int i = 0; i < numBH; i+=2) {
            this.setXIn(1);
            this.setYIn(1);
            this.setXOut(1);
            this.setYOut(1);
            Locator b1 = new BlackHole(xin, yin, xout, yout);
            Locator b2 = new BlackHole(xin, yin, xout, yout);
            board.getPlanet(yin, xin).setHolePl((BlackHole) b1);
            board.getPlanet(yout, xout).setHolePl((BlackHole) b2);
            board.addBlackHole(i, (BlackHole) b1);
            board.addBlackHole(i + 1, (BlackHole) b2);
        }
    }

    /**
     * buat objek warp sesuai jumlah warp
     */
    public void makeAllWG() {
        if(numWG%2!=0){
             numWG++;
        }
        for (int i = 0; i < numWG; i+=2) {
            //warper di genap
            this.setXIn(0);
            this.setYIn(0);
            this.setXOut(0);
            this.setYOut(0);
            Locator w1 = new SpaceWarpGate(xin, yin, xout, yout);
            Locator w2 = new SpaceWarpGate(xin, yin, xout, yout);
            board.getPlanet(yin, xin).setWarpPl((SpaceWarpGate) w1);
            board.getPlanet(yout, xout).setWarpPl((SpaceWarpGate) w2);
            board.addWarper(i, (SpaceWarpGate) w1);
            board.addWarper(i + 1, (SpaceWarpGate) w2);
        }
    }

    /**
     * buat objek player nya
     */
    public void makeAllAs() {
        for (int i = 0; i < 2; i++) {
            Astronot as = new Astronot("astronot" + (i+1) + "");
            board.addAstronot(i, as);
        }
    }

    public boolean endGame(Astronot as, int endPoint) {
        //terjadi menang jika salah satu player sudah mencapai titik 100
        if (as.getLocateAs() == endPoint) {
            return true;

        } else {
            return false;
        }

    }

    /**
     * print board buat text mode console
     */
    public void print() {
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                //player diwakilkan dengan X dan Y
                //kotak kosong dengan Φ
                //black hole dengan β
                //warp dengan ω
                if (board.getPlanet(i, j).getWarpPl() != null) {
                    System.out.print("ω ");
                } else if (board.getPlanet(i, j).getHolePl() != null) {
                    System.out.print("β ");
                } else if (board.getAsAt(0).getLocateAs() == board.getPlanet(i, j).getLocatePl()) {
                    System.out.print("X ");
                } else if (board.getPlanet(i, j).getLocatePl() == board.getAsAt(1).getLocateAs()) {
                    System.out.print("Y ");
                } else {
                    System.out.print("Φ ");
                }
            }
            System.out.println("");
        }
    }

    /**
     * method play 2 player human
     */
    public void play() {
        //inisiasi semua objek dulu
        this.makeAllAs();
        this.makeAllBH();
        this.makeAllWG();
        
        
        boolean lanjut = true;
        //genap player 1,ganjil player 2, penentuan player 1 dan player 2 sebelum inisiasi objek
        //ngga bakal ada player 2 jalan lbh dulu dari player 1
        //penentuan siapa player 1 sebelum permainan di mulai utk di bagian visualisasi
        int giliran = 0;
        while (lanjut) {
            int dadu = board.dice.roll();

            do {
                //kalo gerak out of bounds / melebihi 100
                if (board.getAsAt(giliran).getLocateAs() + dadu > 100) {
                    //System.out.println("JLEGUR ANJING");
                    board.getAsAt(giliran).moveBackward(dadu);
                    board.getAsAt(giliran).setLocateAs();
                } else {
                    for (int i = 0; i < dadu; i++) {
                        board.getAsAt(giliran).move();
                        board.getAsAt(giliran).setLocateAs();
                    }
                }
                if(dadu==6){
                    System.out.println("Lucky 6!!!");
                }
                //kocok lagi
                dadu = board.dice.roll();
                
                //kalo dapet dadu 6 jalan lagi
            } while (dadu == 6);
            
            //System.out.println("x:"+board.getAsAt(giliran).getXAs()+" y:"+board.getAsAt(giliran).getYAs());
            
            if (board.getPlanet(board.getAsAt(giliran).getYAs(), board.getAsAt(giliran).getXAs()).getHolePl()!=null) {
                //kalo nginjek blackhole ubah posisi player ke xout yout
               
                int xawal = board.getAsAt(giliran).getXAs();
                int yawal = board.getAsAt(giliran).getYAs();
                //simpen dulu posisi nya di variabel agar tidak terlalu rujid
                int xout = board.getPlanet(yawal, xawal).getHolePl().getxOutLc();
                int yout = board.getPlanet(yawal, xawal).getHolePl().getyOutLc();
                //ubah posisi player
                board.getAsAt(giliran).setXAs(xout);
                board.getAsAt(giliran).setYAs(yout);
                //update locate player
                //board.getPlanet(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs()).setLocatePl(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs());
                board.getAsAt(giliran).setLocateAs();
                System.out.println(board.getAsAt(giliran).getName()+"kena blackhole ke posisi:"+board.getAsAt(giliran).getLocateAs());
            }
            
            if (board.getPlanet(board.getAsAt(giliran).getYAs(), board.getAsAt(giliran).getXAs()).getWarpPl() != null) {
                //kalo nginjek warp ubah posisi player ke xout yout
                
                int xawal = board.getAsAt(giliran).getXAs();
                int yawal = board.getAsAt(giliran).getYAs();
                //simpen dulu posisi nya di variabel agar tidak terlalu rujid
                int xout = board.getPlanet(yawal, xawal).getWarpPl().getxOutLc();
                int yout = board.getPlanet(yawal, xawal).getWarpPl().getyOutLc();
                //ubah posisi player
                board.getAsAt(giliran).setXAs(xout);
                board.getAsAt(giliran).setYAs(yout);
                //update locate player
                //board.getPlanet(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs()).setLocatePl(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs());
                board.getAsAt(giliran).setLocateAs();
                System.out.println(board.getAsAt(giliran).getName()+"kena warp ke posisi:"+board.getAsAt(giliran).getLocateAs());
            }
            
            //kalo nginjek lawan
            if (giliran == 0) {
                if (board.getAsAt(giliran).getLocateAs() == board.getAsAt(giliran + 1).getLocateAs()) {
                    board.getAsAt(1).setXAs(0);
                    board.getAsAt(1).setYAs(9);
                    board.getAsAt(1).setLocateAs();
                    System.out.println("mampus keinjek!!!");
                }
            }
            if (giliran == 1) {
                if (board.getAsAt(giliran).getLocateAs() == board.getAsAt(0).getLocateAs()) {
                    board.getAsAt(0).setXAs(0);
                    board.getAsAt(0).setYAs(9);
                    board.getAsAt(0).setLocateAs();
                    System.out.println("mampus keinjek!!!");
                }
            }
            //System.out.println(board.getAsAt(giliran).getName() + " pada posisi:" + board.getAsAt(giliran).getLocateAs());
            System.out.println(board.getAsAt(0).getName()+" pada posisi:"+board.getAsAt(0).getLocateAs());
            System.out.println(board.getAsAt(1).getName()+" pada posisi:"+board.getAsAt(1).getLocateAs());
            this.print();

            //end game
            if (this.endGame(board.getAsAt(giliran), 100)) {
                lanjut = false;
                if (giliran == 0) {
                    System.out.println("Player 1 win");
                } else {
                    System.out.println("Player 2 win");
                }
            } else {
                giliran++;
                if (giliran ==2) {
                    giliran = 0;
                }
            }

        //tampilkan siapa yg menang
        }

    }
}

File: php/InterstellarCompVsComp.php
<?php
	class InterstellarCompVsComp{
		private $random;
		private $numBH;
		private $numWG;
		private $board;
		
		private $xin=0;
		private $xout=0;
		private $yin=0;
		private $yout=0;
		
		public function __construct(){
			$this->numBH=rand(2,4);
			if($this->numBH%2==1){
				$this->numBH++;
			}
			$this->numWG=rand(2,4);
			if($this->numWG%2==1){
				$this->numWG++;
			}
			$this->board=new Board(6,4,4,2);
		}
		
		public function set_X_In($isGanjil){
			$x=rand(2,4)+1;
			$x+=$isGanjil;
			if($x%2==0){
				$x++;
			}
			$this->xin=$x;
		}
		
		public function set_X_Out($isGanjil){
			$x=rand(2,4)+1;
			$x+=$isGanjil;
			if($x%2==0){
				$x++;
			}
			$this->xout=$x;
		}
		
		public function set_Y_In($isGanjil){
			$x=rand(2,4)+1;
			$x+=$isGanjil;
			if($x%2==0){
				$x++;
			}
			$this->yout=$x;
		}
		public function set_Y_Out($isGanjil){
			$x=rand(2,4)+1;
			$x+=$isGanjil;
			if($x%2==0){
				$x++;
			}
			$this->yout=$x;
		}
		
		public function make_All_BH(){
			for($i=0;$i<$this->numBH;$i++){
				$this->set_X_In(1);
				$this->set_Y_In(1);
				$this->set_X_Out(1);
				$this->set_Y_Out(1);
				$b1=new BlackHole($this->xin,$this->yin,$this->xout,$this->yout);
				$b2=new BlackHole($this->xin,$this->yin,$this->xout,$this->yout);
				$this->board->get_Planet($this->yin,$this->xin)->set_Hole_Pl($b1);
				$this->board->get_Planet($this->yout,$this->xout)->set_Hole_Pl($b2);
				$this->board->add_Black_Hole($i,$b1);	
				$this->board->add_Black_Hole($i,$b2);
			}
		}
		
		public function make_All_WG(){
			for($i=0;$i<$this->numWG;$i++){
				$this->set_X_In(0);
				$this->set_Y_In(0);
				$this->set_X_Out(0);
				$this->set_Y_Out(0);
				$w1=new SpaceWarpGate($this->xin,$this->yin,$this->xout,$this->yout);
				$w2=new SpaceWarpGate($this->xin,$this->yin,$this->xout,$this->yout);
				$this->board->get_Planet($this->yin,$this->xin)->set_Warp_Pl($w1);
				$this->board->get_Planet($this->yout,$this->xout)->set_Warp_Pl($w2);
				$this->board->add_Warper($i,$w1);	
				$this->board->add_Warper($i,$w2);
			}
		}
		
		public function make_All_As(){
			for($i=0;$i<2;$i++){
				$i++;
				$as=new Astronot("astronot ".$i);
				$i--;
				$this->board->add_Astronot($i,$as);
			}
		}
		
		public function end_Game($as,$endPoint){
			if($as->get_Locate_As()==$endPoint){
				return true;
			}
			else{
				return false;
			}
		}
		
		public function print_Board(){
			for($i=0;$i<10;$i++){
				for($j=0;$j<10;$j++){
					if($this->board->get_Planet($i,$j)->get_Warp_Pl()!=null){
						echo "[ ω ] ";
					}
					else if($this->board->get_Planet($i,$j)->get_Hole_Pl()!=null){
						echo "[ β ] ";
					}
					else if($this->board->get_As_At(0)->get_Locate_As()==$this->board->get_Planet($i,$j)->get_Locate_Pl()){
						echo "[ X ] ";
					}
					else if($this->board->get_As_At(1)->get_Locate_As()==$this->board->get_Planet($i,$j)->get_Locate_Pl()){
						echo "[ Y ] ";
					}
					else{
						echo "[ Φ ] ";
					}
					
				}
				echo "<br>";
					for ($k = 0; $k < 10; $k++) {
					if ($i % 2 == 0) {
						$indexGenap = ((9 - $i) * 10) + ((9 - $k) + 1);
						echo "  |" . " $indexGenap  " ."  | ";
					} else {
						$indexGanjil = ((9 - $i) * 10) + ($k + 1);
						if ($indexGanjil < 10) {
							echo "  |0" . " $indexGanjil  " . "  | ";
						} else {
							echo "  |" . " $indexGanjil  " . "  | ";
						}
					}
				}
				echo "<br>";
			}
		}
		
		public function play(){
			$this->make_All_As();
			$this->make_All_BH();
			$this->make_All_WG();
			$dadu_Main=0;
			$lanjut=true;
			$giliran=0;
			
			while($lanjut){
				echo "<br>";
				//roll dadu
				$dadu_Main=$this->board->roll_The_Dice();
				echo $this->board->get_As_At($giliran)->get_Name()." mendapat dadu:".$dadu_Main;
				echo " ";
				if($this->board->get_As_At($giliran)->get_Locate_As()+$dadu_Main>100){
					$this->board->get_As_At($giliran)->move_Backward($dadu_Main);
					$this->board->get_As_At($giliran)->set_Locate_As();
				}
				else{
					for($i=0;$i<$dadu_Main;$i++){
						$this->board->get_As_At($giliran)->move();
						$this->board->get_As_At($giliran)->set_Locate_As();
					}
				}
				
				if($dadu_Main==6){
					echo "Lucky 6!!"."<br>";
					$dadu_Main=$this->board->roll_The_Dice();
					echo $this->board->get_As_At($giliran)->get_Name()." mendapat dadu:".$dadu_Main;
					if($this->board->get_As_At($giliran)->get_Locate_As()+$dadu_Main>100){
						$this->board->get_As_At($giliran)->move_Backward($dadu_Main);
						$this->board->get_As_At($giliran)->set_Locate_As();
					}
					else{
						for($i=0;$i<$dadu_Main;$i++){
							$this->board->get_As_At($giliran)->move();
							$this->board->get_As_At($giliran)->set_Locate_As();
						}
					}
				}
				
				if ($this->board->get_Planet($this->board->get_As_At($giliran)->get_Y_As(), $this->board->get_As_At($giliran)->get_X_As())->get_Hole_Pl()!=null) {
					//kalo nginjek blackhole ubah posisi player ke xout yout
					echo $this->board->get_As_At($giliran)->get_Name()."kena blackhole"."<br>";
					$xawal = $this->board->get_As_At($giliran)->get_X_As();
					$yawal = $this->board->get_As_At($giliran)->get_Y_As();
					//simpen dulu posisi nya di variabel agar tidak terlalu rujid
					$xout = $this->board->get_Planet($yawal, $xawal)->get_Hole_Pl()->get_x_Out_Lc();
					$yout = $this->board->get_Planet($yawal, $xawal)->get_Hole_Pl()->get_y_Out_Lc();
					//ubah posisi player
					$this->board->get_As_At($giliran)->set_X_As($xout);
					$this->board->get_As_At($giliran)->set_Y_As($yout);
					//update locate player
					//board.getPlanet(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs()).setLocatePl(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs());
					$this->board->get_As_At($giliran)->set_Locate_As();
				}
				
				if ($this->board->get_Planet($this->board->get_As_At($giliran)->get_Y_As(), $this->board->get_As_At($giliran)->get_X_As())->get_Warp_Pl()!=null) {
					//kalo nginjek warp ubah posisi player ke xout yout
					echo $this->board->get_As_At($giliran)->get_Name()."kena warp"."<br>";
					$xawal = $this->board->get_As_At($giliran)->get_X_As();
					$yawal = $this->board->get_As_At($giliran)->get_Y_As();
					//simpen dulu posisi nya di variabel agar tidak terlalu rujid
					$xout = $this->board->get_Planet($yawal, $xawal)->get_Warp_Pl()->get_x_Out_Lc();
					$yout = $this->board->get_Planet($yawal, $xawal)->get_Warp_Pl()->get_y_Out_Lc();
					//ubah posisi player
					$this->board->get_As_At($giliran)->set_X_As($xout);
					$this->board->get_As_At($giliran)->set_Y_As($yout);
					//update locate player
					//board.getPlanet(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs()).setLocatePl(board.getAsAt(0).getYAs(), board.getAsAt(0).getXAs());
					$this->board->get_As_At($giliran)->set_Locate_As();
				}
				
				//kalo nginjek lawan
				if ($giliran == 0) {
					if ($this->board->get_As_At($giliran)->get_Locate_As() == $this->board->get_As_At($giliran + 1)->get_Locate_As()) {
						$this->board->get_As_At(1)->set_X_As(0);
						$this->board->get_As_At(1)->set_Y_As(9);
						$this->board->get_As_At(1)->set_Locate_As();
						echo "mampus keinjek!!!"."<br>";
					}
				}
				if ($giliran == 1) {
					if ($this->board->get_As_At($giliran)->get_Locate_As() == $this->board->get_As_At(0)->get_Locate_As())  {
						$this->board->get_As_At(0)->set_X_As(0);
						$this->board->get_As_At(0)->set_Y_As(9);
						$this->board->get_As_At(0)->set_Locate_As();
						echo "mampus keinjek!!!"."<br>";
					}
				}
				
				if($this->board->get_As_At(0)->get_Locate_As()+1==101){
					
					echo "<br>";
					echo $this->board->get_As_At(0)->get_Name()." pada posisi:".($this->board->get_As_At(0)->get_Locate_As());
					echo " ___ ";
					echo $this->board->get_As_At(1)->get_Name()." pada posisi:".($this->board->get_As_At(1)->get_Locate_As()+1);
					echo "<br>";
				}
				else if($this->board->get_As_At(1)->get_Locate_As()+1==101){
					echo "<br>";
					echo $this->board->get_As_At(0)->get_Name()." pada posisi:".($this->board->get_As_At(0)->get_Locate_As()+1);
					echo " ___ ";
					echo $this->board->get_As_At(1)->get_Name()." pada posisi:".($this->board->get_As_At(1)->get_Locate_As());
					echo "<br>";
				}
				else{
					echo "<br>";
					echo $this->board->get_As_At(0)->get_Name()." pada posisi:".($this->board->get_As_At(0)->get_Locate_As()+1);
					echo " ___ ";
					echo $this->board->get_As_At(1)->get_Name()." pada posisi:".($this->board->get_As_At(1)->get_Locate_As()+1);
					echo "<br>";
				}
				$this->print_Board();
				echo "<br>";
				
				if ($this->end_Game($this->board->get_As_At($giliran), 100)) {
					$lanjut = false;
					if ($giliran == 0) {
						echo "Player 1 win";
						echo "<br>";
					} else {
						echo "Player 2 win";
						echo "<br>";
					}
				} else {
					$giliran++;
					if ($giliran ==2) {
						$giliran = 0;
					}
				}
				
			}
			
		}
		
	}
?>
